panic when passing function to t.Run()

Hey folks,

I just upgraded to golangci-lint v1.44.0 which contains thelper v0.5.0. This version of thelper 
panics when running it against our repository (ClusterAPI). The previous version (v0.4.0) didn't.

The code block in ClusterAPI that triggers the panic seems to be:
```go
t.Run("for Cluster", customDefaultValidateTest(ctx, c, webhook))
```
https://github.com/kubernetes-sigs/cluster-api/blob/82ab810646f11ec7c96a1543697264deaa3884ec/internal/webhooks/cluster_test.go#L352

thelper panics because `f.Obj` is nil here: 
```go
		funObjDecl, ok := f.Obj.Decl.(*ast.FuncDecl)
```
https://github.com/kulti/thelper/blob/master/pkg/analyzer/analyzer.go#L415

# How to reproduce

Clone cluster-api (https://github.com/kubernetes-sigs/cluster-api). Checkout commit `4c4b47d`

cd into the repo and run:
```bash
/thelper ./internal/webhooks/...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x129572d]

goroutine 6016 [running]:
github.com/kulti/thelper/pkg/analyzer.unwrapTestingFunctionBuilding(0x12c0700, {0x1373088, 0xc0014aa7c0}, {0x13729a0, 0xc002ddf9e0})
        /Users/buringerst/code/src/github.com/kulti/thelper/pkg/analyzer/analyzer.go:415 +0x6d
github.com/kulti/thelper/pkg/analyzer.extractSubtestExp(0xc002f03520, 0xc0014aa800, {0x137ddf8, 0xc00067dbd0}, {0x13729a0, 0xc002ddf9e0})
        /Users/buringerst/code/src/github.com/kulti/thelper/pkg/analyzer/analyzer.go:394 +0xe5
github.com/kulti/thelper/pkg/analyzer.thelper.run.func1({0x1372158, 0xc0014aa800})
        /Users/buringerst/code/src/github.com/kulti/thelper/pkg/analyzer/analyzer.go:147 +0xb5
golang.org/x/tools/go/ast/inspector.(*Inspector).Preorder(0xc001593878, {0xc00261b650, 0x1520700, 0x1}, 0xc00261b950)
        /Users/buringerst/code/pkg/mod/golang.org/x/tools@v0.0.0-20201114224030-61ea331ec02b/go/ast/inspector/inspector.go:77 +0x86
github.com/kulti/thelper/pkg/analyzer.thelper.run({0x12cf480}, 0xc002f03520)
        /Users/buringerst/code/src/github.com/kulti/thelper/pkg/analyzer/analyzer.go:133 +0x40b
golang.org/x/tools/go/analysis/internal/checker.(*action).execOnce(0xc001252000)
        /Users/buringerst/code/pkg/mod/golang.org/x/tools@v0.0.0-20201114224030-61ea331ec02b/go/analysis/internal/checker/checker.go:691 +0xa1e
sync.(*Once).doSlow(0x125ec29, 0xc00014c1c0)
        /usr/local/Cellar/go/1.17.2/libexec/src/sync/once.go:68 +0xd2
sync.(*Once).Do(...)
        /usr/local/Cellar/go/1.17.2/libexec/src/sync/once.go:59
golang.org/x/tools/go/analysis/internal/checker.(*action).exec(0xc001592fa8)
        /Users/buringerst/code/pkg/mod/golang.org/x/tools@v0.0.0-20201114224030-61ea331ec02b/go/analysis/internal/checker/checker.go:579 +0x3d
golang.org/x/tools/go/analysis/internal/checker.execAll.func1(0x0)
        /Users/buringerst/code/pkg/mod/golang.org/x/tools@v0.0.0-20201114224030-61ea331ec02b/go/analysis/internal/checker/checker.go:567 +0x25
created by golang.org/x/tools/go/analysis/internal/checker.execAll
        /Users/buringerst/code/pkg/mod/golang.org/x/tools@v0.0.0-20201114224030-61ea331ec02b/go/analysis/internal/checker/checker.go:573 +0x168
```

# Workaround

When I split:
```go
t.Run("for Cluster", customDefaultValidateTest(ctx, c, webhook))
```
into two statements:
```go
testFunc := customDefaultValidateTest(ctx, c, webhook)
t.Run("for Cluster", testFunc)
```
everything works as expected.

Related ClusterAPI PR to upgrade to golangci-lint v1.44 / thelper v0.5.0 (https://github.com/kubernetes-sigs/cluster-api/pull/6014)

--- a/pkg/analyzer/analyzer.go
+++ b/pkg/analyzer/analyzer.go
@@ -420,7 +420,7 @@ func unwrapTestingFunctionBuilding(pass *analysis.Pass, expr ast.Expr, testFuncT
 		funcDecl.Body = funObjDecl.Body
 		funcDecl.Type = funObjDecl.Type
 	case *ast.SelectorExpr:
-		fd := findSelectroDeclaration(pass, f)
+		fd := findSelectorDeclaration(pass, f)
 		if fd == nil {
 			return nil
 		}
@@ -499,8 +499,8 @@ func isExprHasType(pass *analysis.Pass, expr ast.Expr, expType types.Type) bool 
 	return types.Identical(typeInfo.Type, expType)
 }
 
-// findSelectroDeclaration returns function declaration called by selectro expression.
-func findSelectroDeclaration(pass *analysis.Pass, expr *ast.SelectorExpr) *ast.FuncDecl {
+// findSelectorDeclaration returns function declaration called by selector expression.
+func findSelectorDeclaration(pass *analysis.Pass, expr *ast.SelectorExpr) *ast.FuncDecl {
 	xsel, ok := pass.TypesInfo.Selections[expr]
 	if !ok {
 		return nil

--- a/pkg/analyzer/analyzer.go
+++ b/pkg/analyzer/analyzer.go
@@ -412,8 +412,8 @@ func unwrapTestingFunctionBuilding(pass *analysis.Pass, expr ast.Expr, testFuncT
 		funcDecl.Body = f.Body
 		funcDecl.Type = f.Type
 	case *ast.Ident:
-		funObjDecl, ok := f.Obj.Decl.(*ast.FuncDecl)
-		if !ok {
+		funObjDecl := findFunctionDeclaration(pass, f)
+		if funObjDecl == nil {
 			return nil
 		}
 
@@ -527,6 +527,35 @@ func findSelectroDeclaration(pass *analysis.Pass, expr *ast.SelectorExpr) *ast.F
 				if fd.Name.Name == expr.Sel.Name {
 					return fd
 				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// findFunctionDeclaration returns function declaration called by identity.
+func findFunctionDeclaration(pass *analysis.Pass, ident *ast.Ident) *ast.FuncDecl {
+	if ident.Obj != nil {
+		if funObjDecl, ok := ident.Obj.Decl.(*ast.FuncDecl); ok {
+			return funObjDecl
+		}
+	}
+
+	obj := pass.TypesInfo.ObjectOf(ident)
+	if obj == nil {
+		return nil
+	}
+
+	for _, file := range pass.Files {
+		for _, decl := range file.Decls {
+			funcDecl, ok := decl.(*ast.FuncDecl)
+			if !ok {
+				continue
+			}
+
+			if funcDecl.Name.Pos() == obj.Pos() {
+				return funcDecl
 			}
 		}
 	}

--- a/pkg/analyzer/testdata/src/t/s.go
+++ b/pkg/analyzer/testdata/src/t/s.go
@@ -36,6 +36,7 @@ func (sh subhelper) anotherCheck(t *testing.T) {} // want "test helper function 
 func TestSubtestWithBuilder(t *testing.T) {
 	t.Run("sub1", subtestBuilder("first"))
 	t.Run("sub2", subtestBuilder("second"))
+	t.Run("sub_from_other_file", subtestBuilderAnotherFile())
 	t.Run("anon", func() func(t *testing.T) { return func(t *testing.T) {} }())
 
 	fixture := subtestFixture{}

--- a/pkg/analyzer/testdata/src/t/ss.go
+++ b/pkg/analyzer/testdata/src/t/ss.go
@@ -1,0 +1,7 @@
+package t
+
+import "testing"
+
+func subtestBuilderAnotherFile() func(*testing.T) {
+	return func(t *testing.T) {}
+}

